Fix #45 - Multiple familes in ped file

stevenhwu opened the pull request:
- Fix #45 <- test (https://help.github.com/articles/closing-issues-via-commit-messages/)
- Merge family ID and individual ID in `ped.h` to make id unique. Use an "unique/unusual/obscure" delimiter, `_F@I_`, to concatenate these two IDs, then everything will be unique in `table_`. The `id()` function take care of the rest.
  - A simple delimiter (`_`, `@` ...etc) is avoid to prevent the case that family name is something like `*f1@` or `*f1_`.
- Update related unittests.
# 

Let's assume we have a pedigree looks like the following

```
1       1       0       0       1       NA12891
1       2       0       0       2       NA12892
1       3       1       2       2       NA12878
2       1       0       0       1       NA12900
2       2       0       0       2       NA12901
2       3       1       2       1       NA12902
```

In `ped.h` the `Parse()` function.

The first part relate to `string_table`, we append family id and individual id together in the `child_names` (instead of just individual id). It should looks something like

```
1_F@I_1
1_F@I_2
1_F@I_2
2_F@I_1
2_F@I_2
2_F@I_3
```

We pick a pretty "unique/unusual/obscure" delimiter, in this case `_F@I_`, where `_F` is the suffix for family and `I_` is the prefix for individual. A simple "_" or "@" is avoid to prevent the case that family name is something like `*f1@` or `*f1_`.

The delimiter can be update at `const std::string FAM_IND_DELIM = "_F@I_";`

The second part, where we parse it to `table_` with struct `Member` where everything is `size_t`. The `id()` function automatically take care of that due to the way `child_names` in constructed.

The new output looks like

```
##PEDIGREE=<ID=LB-NA12878:Solexa-135852,Father=GL-1_F@I_1,Mother=GL-1_F@I_2,FatherMR=3.0e-08,MotherMR=3.0e-08>
##PEDIGREE=<ID=LB-NA12891:Solexa-135851,Original=GL-1_F@I_1,OriginalMR=2.0e-08>
##PEDIGREE=<ID=LB-NA12892:Solexa-135853,Original=GL-1_F@I_2,OriginalMR=2.0e-08>
##PEDIGREE=<ID=LB-NA12900:L900,Original=GL-2_F@I_1,OriginalMR=2.0e-08>
##PEDIGREE=<ID=LB-NA12901:L901,Original=GL-2_F@I_2,OriginalMR=2.0e-08>
##PEDIGREE=<ID=LB-NA12902:L902,Father=GL-2_F@I_1,Mother=GL-2_F@I_2,FatherMR=3.0e-08,MotherMR=3.0e-08>
```

stevenhwu committed:
Fix Issue #45 - Multiple familes in ped file

- Merge family ID and individual ID in `ped.h` to make id unique. Use an "unique/unusual/obscure" delimiter, `_F@I_`, to concatenate these two IDs, then everything will be unique in `table_`. The `id()` function take care of the rest.
  - A simple delimiter (`_`, `@` ...etc) is avoid to prevent the case that family name is something like `*f1@` or `*f1_`.
- Update related unittests.

dng-jenkins commented:
Can one of the admins verify this patch?

reedacartwright commented:
Jenkins, okay to test.

--
Reed A. Cartwright, PhD
Barrett Honors Faculty
Assistant Professor of Genomics, Evolution, and Bioinformatics
School of Life Sciences
Human and Comparative Genomics Laboratory
The Biodesign Institute
Arizona State University
==================
Availability: http://links.asu.edu/CartwrightCalendar
Address: The Biodesign Institute, PO Box 875301, Tempe, AZ 85287-5301 USA
Packages: The Biodesign Institute, 1001 S. McAllister Ave, Tempe, AZ
85287-5301 USA
Office: Biodesign A-224A, 1-480-965-9949
Website: http://cartwrig.ht/

On Tue, Oct 4, 2016 at 3:54 PM, Jenkins Bot <notifications@github.com>
wrote:

> Can one of the admins verify this patch?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/denovogear/denovogear/pull/196#issuecomment-251537119>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGOHlpBVxOZ_WZZRSPzIzFwpXDZcT1uks5qwtk4gaJpZM4KOP6f>
> .
>

reedacartwright commented:
I would use probably use '\t' as an internal separator since fields cannot
contain a tab.

--
Reed A. Cartwright, PhD
Barrett Honors Faculty
Assistant Professor of Genomics, Evolution, and Bioinformatics
School of Life Sciences
Human and Comparative Genomics Laboratory
The Biodesign Institute
Arizona State University
==================
Availability: http://links.asu.edu/CartwrightCalendar
Address: The Biodesign Institute, PO Box 875301, Tempe, AZ 85287-5301 USA
Packages: The Biodesign Institute, 1001 S. McAllister Ave, Tempe, AZ
85287-5301 USA
Office: Biodesign A-224A, 1-480-965-9949
Website: http://cartwrig.ht/

On Tue, Oct 4, 2016 at 4:52 PM, Reed A. Cartwright <cartwright@asu.edu>
wrote:

> Jenkins, okay to test.
>
> --
> Reed A. Cartwright, PhD
> Barrett Honors Faculty
> Assistant Professor of Genomics, Evolution, and Bioinformatics
> School of Life Sciences
> Human and Comparative Genomics Laboratory
> The Biodesign Institute
> Arizona State University
> ==================
> Availability: http://links.asu.edu/CartwrightCalendar
> Address: The Biodesign Institute, PO Box 875301, Tempe, AZ 85287-5301 USA
> Packages: The Biodesign Institute, 1001 S. McAllister Ave, Tempe, AZ
> 85287-5301 USA
> Office: Biodesign A-224A, 1-480-965-9949
> Website: http://cartwrig.ht/
>
> On Tue, Oct 4, 2016 at 3:54 PM, Jenkins Bot <notifications@github.com>
> wrote:
>
>> Can one of the admins verify this patch?
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/denovogear/denovogear/pull/196#issuecomment-251537119>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAGOHlpBVxOZ_WZZRSPzIzFwpXDZcT1uks5qwtk4gaJpZM4KOP6f>
>> .
>>
>
>

stevenhwu commented:
Then the output in vcf will look something like
```
##PEDIGREE=<ID=LB-NA12892:Solexa-135853,Original=GL-1  2,OriginalMR=2.0e-08>
                                                     ^^
##PEDIGREE=<ID=LB-NA12900:L900,Original=GL-2    1,OriginalMR=2.0e-08>
                                            ^^^^ each line will have different amount of spaces
```
Which might make `BamTest.cmake.in` hard to create.

OR we can parse internally with `\t`, then output to vcf with something other than `\t`

stevenhwu commented:
btw, can someone explain the coverage report please? @reedacartwright @anderspitman 
I really expect the delta is 0% here.
```
Current coverage is 65.36% (diff: 100%)

>Merging #196 into develop will increase coverage by 8.18%
```

stevenhwu committed:
Update delimiter and unittest

- Internal delimiter is `\t`
- Output delimiter, which is only used if required, is `/`
- Add unittests
- Output warning message for duplicate entries (same family and same individual). Currently the duplicate entry is ignored.
- Assumes most of the cases only one family is in the ped, so `if(is_individual_id_not_unique){...}` looks redundant.

stevenhwu committed:
Merge branch 'develop' into feature/multi_families_in_ped

stevenhwu committed:
Rename `GL(SM)(LB)-` to `GL(SM)(LB)/`

```
```
to
```
```

reedacartwright commented on Tests/Unit/dng/find_mutation.cc:
```diff
@@ -216,7 +216,7 @@ BOOST_FIXTURE_TEST_CASE(test_calculate_mutation_expected, TrioWorkspace) {
 
     BOOST_CHECK_EQUAL("GGxGG>GT", mutation_stats.dnt);
-    BOOST_CHECK_EQUAL("LB-NA12878:Solexa-135852", mutation_stats.dnl);
+    BOOST_CHECK_EQUAL("LB/NA12878:Solexa-135852", mutation_stats.dnl);
```
Looks like running `dng-call` on a vcf still assumes that names are "sample:library" instead of "sample/library".

Can we fix that in this PR too?

stevenhwu replied:
Yes, this is easy in ReadGroup. But are we sure about this?
In read_group.h we got
Change everything to SAMPLE/LIBRARY, include both input and output
Need to match the last `/`, e.g. SM/1/Lib

stevenhwu replied:
Done.

stevenhwu replied:
Done (reply in review).

reedacartwright commented on src/include/dng/io/ped.h:
```diff
@@ -65,23 +70,30 @@ class Pedigree {
+//        }
+//        return output_names_.project<0>(it) - output_names_.begin();
+//    }
+    std::size_t id(const std::string &name, NameContainer &names) const {
```
static function, right?

stevenhwu replied:
Done.

stevenhwu replied:
Done (reply in review).

reedacartwright commented on src/include/dng/io/ped.h:
```diff
@@ -124,15 +136,50 @@ class Pedigree {
         // Build map of unique child names
-        map<string, size_t> child_names;
-        child_names.emplace("", 0);
+        bool is_individual_id_not_unique = false;
```
Add a TODO here for me to look at later. I want to think about this algorithm.

stevenhwu replied:
Add TODO(RC): ****

stevenhwu replied:
Add TODO(RC): ***\*  
(add to review)

reedacartwright commented on src/include/dng/io/ped.h:
```diff
@@ -187,33 +238,45 @@ class Pedigree {
+                    child,
+                    id(string_table[nrow][2], output_order),
+                    id(string_table[nrow][3], output_order),
+                    ParseGender(string_table[nrow][4]),
```
sex

stevenhwu replied:
Done (reply in review).

reedacartwright commented on src/lib/newick.cc:
```diff
@@ -31,7 +31,7 @@
 #include <boost/algorithm/string/trim_all.hpp>
 
-#define DNG_SM_PREFIX "SM-" // define also in pedigree.cc
+#define DNG_SM_PREFIX "SM/" // define also in relationship_graph.cc
```
Is there a good way we can move these definitions to a common header file?

stevenhwu replied:
Introduce `constant.h`

stevenhwu replied:
Introduce constant.h (add review)

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -25,9 +25,9 @@
-#define DNG_LB_PREFIX "LB-"
+#define DNG_GL_PREFIX "GL/"
+#define DNG_SM_PREFIX "SM/" // define also in newick.cc
+#define DNG_LB_PREFIX "LB/"
```
Can we move these to a header file? So we can keep newick.cc in sync with relationship_graph?

stevenhwu replied:
Option 1: If we must use #define, then I suggest create a `constant.h` file.
Option 2: If we don't need to use #define, how about `extern const std::string .....` in *.h

reedacartwright replied:
Using a define allows the compiler to optimize the strings better as "a" "b" becomes "ab":

> String literals placed side-by-side are concatenated at translation phase 6 (after the preprocessor). That is, "Hello,"  " world!" yields the (single) string "Hello, world!". If the two strings have the same encoding prefix (or neither has one), the resulting string will have the same encoding prefix (or no prefix).

http://en.cppreference.com/w/cpp/language/string_literal

reedacartwright reviewed (changes requested):
Are we using the slash to join the library and the sample name as well?

stevenhwu committed:
Change based on review, not done yet

stevenhwu commented:
Accidentally push `e2b3b15`, ignore this commit

stevenhwu committed:
Address review and requested changes

  - Changed everything to `SAMPLE/LIBRARY`, in both `io/ped.h` and `read_group.h`
  - Added `constant.h` for #define DNG_(GL|SM|LB)_PREFIX
  - Resolved merge conflict. Merged from develop - 36bbe9970266389a8a66a5887502fa418e05c45f

stevenhwu reviewed (commented):
Address all requested changes.
Where will this message show up?
Rather useless here.....

stevenhwu reviewed (commented):
Reply all comments in "review". 
Is this helpful?

stevenhwu committed:
Fix merge conflict take 2

- without squash

reedacartwright commented:
Fixed in #220.
